File: order-service/src/main/java/order_service/entity/Order.java
package order_service.entity;

import io.swagger.v3.oas.annotations.media.Schema;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@Entity
@Table(name = "`Order`")
@Data
@AllArgsConstructor
@NoArgsConstructor
@Schema(description = "Order entity representing customer orders in the Healthy Food Bank system")
public class Order {

    @PrePersist
    protected void onCreate() {
        orderPlacedDate = LocalDateTime.now();
        statusUpdatedAt = LocalDateTime.now();
        if (orderStatus == null) {
            orderStatus = "ISSUED";
        }
    }

    @PreUpdate
    protected void onUpdate() {
        statusUpdatedAt = LocalDateTime.now();
    }

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Schema(description = "Unique identifier for the order", example = "1", accessMode = Schema.AccessMode.READ_ONLY)
    private int id;
    
    @Schema(description = "Name/description of the ordered item", example = "Organic Apples", required = true)
    private String orderName;
    
    @Schema(description = "Quantity of the ordered item", example = "5.0", required = true)
    private double orderQuantity;
    
    @Schema(description = "Unit of measurement for the order", example = "kg", required = true)
    private String orderUnit;
    
    @Schema(description = "Total price of the order", example = "14.95", required = true)
    private double orderPrice;
    
    @Schema(description = "Timestamp when the order was placed (ISSUED status)", accessMode = Schema.AccessMode.READ_ONLY)
    private LocalDateTime orderPlacedDate;

    @Schema(description = "Timestamp when the order was scheduled by vendor (ISSUED → SCHEDULED)", accessMode = Schema.AccessMode.READ_ONLY)
    private LocalDateTime scheduledDate;

    @Schema(description = "Timestamp when the order was marked ready for pickup (SCHEDULED → READY)", accessMode = Schema.AccessMode.READ_ONLY)
    private LocalDateTime readyDate;

    @Schema(description = "Timestamp when the order was delivered to customer (READY → DELIVERED)", accessMode = Schema.AccessMode.READ_ONLY)
    private LocalDateTime deliveredDate;

    @Deprecated
    @Schema(description = "Deprecated: Use deliveredDate instead", accessMode = Schema.AccessMode.READ_ONLY)
    private LocalDateTime orderDeliveredDate;

    @Schema(description = "Timestamp when the order status was last updated (for polling notifications)", accessMode = Schema.AccessMode.READ_ONLY)
    private LocalDateTime statusUpdatedAt;
    
    @Schema(description = "ID of the customer who placed this order", example = "123", required = true)
    private Long customerId;

    @Schema(description = "Current status of the order", example = "ISSUED",
            allowableValues = {"ISSUED", "SCHEDULED", "CANCELLED_BY_VENDOR"})
    private String orderStatus;

    @Schema(description = "ID of the product being ordered", example = "1")
    private Integer productId;

    @Schema(description = "ID of the vendor who owns the product", example = "VENDOR001")
    private String vendorId;

    @Schema(description = "Name of the vendor who owns the product", example = "John Doe")
    private String vendorName;

    @Schema(description = "Name of the product being ordered", example = "Organic Apples")
    private String productName;

    @Schema(description = "Name of the customer who placed the order", example = "Jane Smith")
    private String customerName;

    @Schema(description = "Phone number of the customer", example = "+91 9876543210")
    private String customerPhone;

    @Schema(description = "Pickup point selected by the customer", example = "Main Street Hub")
    private String customerPickupPoint;

}

File: healthy_food_bank_flutter/lib/models/order.dart
enum OrderStatus {
  PENDING,
  PROCESSING,
  DELIVERED,
  CANCELLED,
  ISSUED,
  SCHEDULED,
  READY,
  CANCELLED_BY_VENDOR,
}

class Order {
  final int? id;
  final String orderName;
  final int orderQuantity;
  final String? orderUnit;
  final double orderPrice;
  final DateTime? orderPlacedDate;
  final DateTime? scheduledDate;
  final DateTime? readyDate;
  final DateTime? deliveredDate;
  final DateTime? orderDeliveredDate; // Deprecated: kept for backward compatibility
  final DateTime? statusUpdatedAt; // For polling-based notifications
  final int? customerId;
  final OrderStatus status;
  final int? productId;
  final String? vendorId;
  final String? productName;
  final String? customerName;
  final String? customerPhone;
  final String? customerPickupPoint;

  Order({
    this.id,
    required this.orderName,
    required this.orderQuantity,
    this.orderUnit,
    required this.orderPrice,
    this.orderPlacedDate,
    this.scheduledDate,
    this.readyDate,
    this.deliveredDate,
    this.orderDeliveredDate,
    this.statusUpdatedAt,
    this.customerId,
    this.status = OrderStatus.PENDING,
    this.productId,
    this.vendorId,
    this.productName,
    this.customerName,
    this.customerPhone,
    this.customerPickupPoint,
  });

  factory Order.fromJson(Map<String, dynamic> json) {
    return Order(
      id: json['id'],
      orderName: json['orderName'] ?? '',
      orderQuantity: json['orderQuantity'] ?? 0,
      orderUnit: json['orderUnit'],
      orderPrice: (json['orderPrice'] ?? 0).toDouble(),
      orderPlacedDate: json['orderPlacedDate'] != null
          ? DateTime.tryParse(json['orderPlacedDate'].toString())
          : null,
      scheduledDate: json['scheduledDate'] != null
          ? DateTime.tryParse(json['scheduledDate'].toString())
          : null,
      readyDate: json['readyDate'] != null
          ? DateTime.tryParse(json['readyDate'].toString())
          : null,
      deliveredDate: json['deliveredDate'] != null
          ? DateTime.tryParse(json['deliveredDate'].toString())
          : null,
      orderDeliveredDate: json['orderDeliveredDate'] != null
          ? DateTime.tryParse(json['orderDeliveredDate'].toString())
          : null,
      statusUpdatedAt: json['statusUpdatedAt'] != null
          ? DateTime.tryParse(json['statusUpdatedAt'].toString())
          : null,
      customerId: json['customerId'],
      status: _parseStatus(json['orderStatus'] ?? 'PENDING'),
      productId: json['productId'],
      vendorId: json['vendorId']?.toString(),
      productName: json['productName'],
      customerName: json['customerName'],
      customerPhone: json['customerPhone']?.toString(),
      customerPickupPoint: json['customerPickupPoint'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'orderName': orderName,
      'orderQuantity': orderQuantity,
      if (orderUnit != null) 'orderUnit': orderUnit,
      'orderPrice': orderPrice,
      if (productId != null) 'productId': productId,
      if (vendorId != null) 'vendorId': vendorId,
      if (customerName != null) 'customerName': customerName,
      if (customerPhone != null) 'customerPhone': customerPhone,
      if (customerPickupPoint != null) 'customerPickupPoint': customerPickupPoint,
    };
  }

  bool get isActive =>
      status == OrderStatus.PENDING ||
      status == OrderStatus.PROCESSING ||
      status == OrderStatus.ISSUED ||
      status == OrderStatus.SCHEDULED ||
      status == OrderStatus.READY;

  String get statusDisplay {
    switch (status) {
      case OrderStatus.CANCELLED_BY_VENDOR:
        return 'Cancelled by Vendor';
      default:
        return status.name[0] + status.name.substring(1).toLowerCase();
    }
  }

  static OrderStatus _parseStatus(String status) {
    switch (status.toUpperCase()) {
      case 'PROCESSING':
        return OrderStatus.PROCESSING;
      case 'DELIVERED':
        return OrderStatus.DELIVERED;
      case 'CANCELLED':
        return OrderStatus.CANCELLED;
      case 'ISSUED':
        return OrderStatus.ISSUED;
      case 'SCHEDULED':
        return OrderStatus.SCHEDULED;
      case 'READY':
        return OrderStatus.READY;
      case 'CANCELLED_BY_VENDOR':
        return OrderStatus.CANCELLED_BY_VENDOR;
      default:
        return OrderStatus.PENDING;
    }
  }
}
